Add support for STP and Yices2

gussmith23 opened the pull request:
Coauthored by @vcanumalla.

TODO before converting from draft:
- [x] Freeze STP at specific checkout
- [x] Make sure tests pass locally

gussmith23 commented:
Ready for review @sorawee

sorawee commented on rosette/solver/smt/yices-smt2.rkt:
```diff
@@ -0,0 +1,69 @@
+         "../solver.rkt"
+         (prefix-in base/ "base-solver.rkt"))
+
+(provide (rename-out [make-yices-smt2 yices-smt2]) yices-smt2? yices-smt2-available?)
```
Why is this called `yices-smt2` rather than e.g. `yices2` (or `yices`)?

gussmith23 replied:
See here: https://github.com/SRI-CSL/yices2/blob/e27cf308cffb0ecc6cc7165c10e81ca65bc303b3/doc/NOTES#L1182

Apparently they make different binaries for each frontend (rather than changing the frontend with a flag, as other solvers do)

emina replied:
I think this should be called `yices`, for consistency with other Rosette solvers.

sorawee commented:
Please also add [documentation here](https://github.com/emina/rosette/blob/master/rosette/guide/scribble/datatypes/solvers%2Bsolutions.scrbl#L281). And please do that for cvc5 and Bitwuzla as well. I overlooked this in #260, which was my bad.

If I understand correctly, on Mac, the `brew` installation of `stp` doesn't work, and it must be compiled from source. If that's the case, it would be good to add the version requirement _somewhere_. Perhaps the documentation might be a good place? I'm open to other possibilities too if you have an idea.

gussmith23 commented:
> Please also add [documentation here](https://github.com/emina/rosette/blob/master/rosette/guide/scribble/datatypes/solvers%2Bsolutions.scrbl#L281). And please do that for cvc5 and Bitwuzla as well (perhaps as a separate PR). I overlooked this in #260, which was my bad.

Done. I added docs for all four solvers in this PR, hope you don't mind. LMK if you want me to move the CVC5/Bitwuzla stuff to a different PR, but it's mostly copy-paste and find-and-replace.

> If I understand correctly, on Mac, the `brew` installation of `stp` doesn't work, and it must be compiled from source. If that's the case, it would be good to add the version requirement _somewhere_. Perhaps the documentation might be a good place? I'm open to other possibilities too if you have an idea.

I added a specific note in the STP docs about this!

gussmith23 commented:
will fix tomorrow!

On Wed, Dec 13, 2023, 5:36 PM Emina Torlak ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In rosette/solver/smt/yices-smt2.rkt
> <https://github.com/emina/rosette/pull/273#discussion_r1426068303>:
>
> > @@ -0,0 +1,69 @@
> +#lang racket
> +
> +(require racket/runtime-path
> +         "server.rkt" "env.rkt"
> +         "../solver.rkt"
> +         (prefix-in base/ "base-solver.rkt"))
> +
> +(provide (rename-out [make-yices-smt2 yices-smt2]) yices-smt2? yices-smt2-available?)
>
> I think this should be called yices, for consistency with other Rosette
> solvers.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/emina/rosette/pull/273#discussion_r1426068303>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJFZAMO2YRU4RNZHFKLRNTYJJJZ7AVCNFSM6AAAAABATONJG2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOOBQG43TMNBTG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sorawee commented on .github/workflows/tests.yml:
```diff
@@ -57,7 +59,33 @@ jobs:
+          ./configure --prefix=$PWD/out/ &&
+          make install &&
+          popd &&
+          cp yices2/out/bin/yices-smt2 bin/yices-smt2
```
After compilation, does Yices produce an executable named `yices-smt2` or `yices_smt2`. This line seems to indicate that it produces `yices-smt2`, but the documentation says that it's `yices_smt2`.

gussmith23 replied:
It's strange. I explain it in the documentation, but:
- after `make` it produces `yices_smt2`
- after `make install` it produces `yices-smt2`
quite frustrating. I documented this fact and also documented the fact that the executable that comes out of just `make` can simply be renamed. I think using `yices-smt2` is the correct option b/c this is the final name of the executable, and the name that's used when Yices is installed e.g. via `apt`.

sorawee commented on rosette/guide/scribble/datatypes/solvers+solutions.scrbl:
```diff
@@ -279,6 +279,145 @@ Returns true if the Boolector solver is available for use (i.e., Rosette can loc
+}
+
+@defproc[(yices-available?) boolean?]{
+Returns true if the Yices2 solver is available for use (i.e., Rosette can locate a @tt{yices} binary).
```
This should be `yices-smt2` (or `yices_smt2`, depending on the executable name that the build process actually produces).

gussmith23 replied:
Great catch, fixed

gussmith23 commented:
Ready for another review, tests are running now. 

BTW: Yices2 tests didn't actually run before, because of the confusion about binary names. (you can see that CI ran STP tests and then finished!) Will have to explicitly check for the Yices2 tests in the logs once CI completes.

sorawee commented on rosette/guide/scribble/datatypes/solvers+solutions.scrbl:
```diff
@@ -279,6 +279,145 @@ Returns true if the Boolector solver is available for use (i.e., Rosette can loc
+
+@defproc[(bitwuzla-available?) boolean?]{
+Returns true if the Bitwuzla solver is available for use (i.e., Rosette can locate a @tt{bitwuzla} binary).
+If this returns @racket[#f], @racket[(bitwuzla)] will not succeed
```
To make these hyperlinked correctly, you need to add `rosette/solver/smt/bitwuzla`, `rosette/solver/smt/cvc5`, etc. under `for-label` at the top of this file.

gussmith23 replied:
Done

sorawee commented on rosette/guide/scribble/datatypes/solvers+solutions.scrbl:
```diff
@@ -279,6 +283,145 @@ Returns true if the Boolector solver is available for use (i.e., Rosette can loc
+Rosette specifically uses the @tt{yices-smt2} executable, which is the Yices2
+solver with its SMTLIB2 frontend enabled.
+Note that just building (without installing) Yices2 will produce an executable
+named @tt{yices_smt2}. This executable can safely be renamed or symlinked to
```
I would add a sentence before "This executable can safely be renamed" to indicate that the installation step will give the correct executable name. Then, say "Alternatively, the built (but not installed) `yices_smt2` executable can safely be renamed ..."

gussmith23 replied:
Done!

sorawee commented:
I'm sorry. I missed that you also need to change `declare-exporting` at the top. Right now the links are red underlined because of that.

<img width="366" alt="Screenshot 2023-12-14 at 12 26 36 PM" src="https://github.com/emina/rosette/assets/9099577/9405357f-043b-41b0-ba9e-66a22f552483">

gussmith23 commented:
Done. Sorry, could have checked that locally myself. Looks fine now, locally.

sorawee commented:
This looks good to me, thanks!
